File: HashMap/Subarray sums divisible by k/SubarraySum.java
import java.util.HashMap;

class Solution {
    public int subarraysDivByK(int[] arr, int k){
        int count =0; //store the num of subarrays divisible by k
        int rem = 0; //to store values of reminder
        int sum = 0;

        HashMap<Integer, Integer> m = new HashMap<>();
        //initialize rem 0 with frequency 1
        m.put(0,1);

        for(int i=0; i<arr.length; i++){
            //calculate the sum of current element
            sum += arr[i];

            //calculate the reminder of the surrent sum
            rem = sum % k;

            //if rem is -ve, make it +ve
            if(rem < 0){
                rem += k;
            }

            //if the rem alredy exist in the map
            if(m.containsKey(rem)){
                count += m.get(rem);  //add the frequency to the count
                m.put(rem, m.get(rem)+1);  //update the frequency
            } else{
                m.put(rem, 1); //if not, then initialize it with 1
            }
        }
        return count;
    }
}

File: HashMap/Subarray sums divisible by k/SubarraySum.cpp
#include <vector>
#include <unordered_map>

class Solution{
public: 
    int subarraysDivByK(vector<int> &arr, int k){
        int count = 0; //count the number of subarrays with sum divisible by k 
        int rem = 0; //store the value of reminder
        int sum = 0; 

        unordered_map<int, int> m;
        //initialize the map with 0 -> 1 to handle subarrays starting from index 0
        m[0] = 1;

        for(int i=0; i<arr.size(); i++){
            //update the sum with current element
            sum += arr[i];

            //calculate the reminder for that sum
            rem = sum % k;

            //if rem is -ve, adjust it to +ve value b/w 0 and k-1
            if(rem < 0){
                rem += k;
            }

            //check if rem has been seen before
            if(m.find(rem) != m.end()){
                count += m[rem];  //add the freq of the rem to the count
                m[rem]++;   //update the freq
            } else{
                m[rem] = 1; //otherwise, initialize it to 1
            }

        }
        
        return count; 
    }
};